### PR TITLE
add freeze, unfreeze methods with experimental tag

### PR DESCRIPTION
**Pull request by nithinraok**
# What does this PR do ?

Add `freeze()` / `unfreeze()` methods to the ASR TransformerEncoder and mark the class as experimental.

**Collection**: ASR

# Changelog

- Add `freeze()` method to `TransformerEncoder`: sets `requires_grad=False` on all parameters and switches the module to eval mode.
- Add `unfreeze(partial: bool = False)` method to `TransformerEncoder: sets requires_grad=True` on all parameters; switches to train mode unless partial=True.
- Decorate `TransformerEncoder` with `@experimental` from `nemo.utils.decorators` to signal API stability is not yet guaranteed.

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 1]
```diff
@@ -239,3 +242,14 @@ def forward(self, audio_signal, length):
...
+    def freeze(self):
+        for p in self.parameters():
+            p.requires_grad = False
+        self.eval()
+
+    def unfreeze(self, partial: bool = False):
+        for p in self.parameters():
+            p.requires_grad = True
+        if not partial:
+            self.train()
```
[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/collections/asr/modules/transformer_encoder.py]: Thanks. Added necessary functionality to the encoder.

[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/collections/asr/modules/transformer_encoder.py]: Added test.

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 2]
```diff
@@ -239,3 +242,27 @@ def forward(self, audio_signal, length):
         x = self.final_norm(x)
         x = x.transpose(1, 2)  # (B, T, D) -> (B, D, T)
         return x, length
+
+    def freeze(self):
```
Comment (pzelasko) [inline nemo/collections/asr/modules/transformer_encoder.py]: nitpick but maybe this should be either a mixin, or a standalone module `ModelFreezer` or free function `freeze(module) / unfreeze(module, partial)`? the logic seems quite generic, not sure if it should be included in class definition. this class could just do

```
def freeze(self):
    return freeze(self)
def unfreeze(self, partial):
    return unfreeze(self, partial=partial)
```

**Review by pzelasko — DISMISSED** [review]

**Comment by pzelasko** [comment]
/ok to test [e1f0bc4](https://github.com/NVIDIA-NeMo/NeMo/pull/15729/commits/e1f0bc437dcf6b2c32c7441ecefb9bd7518ea446)

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 2 above)
Comment (nithinraok) [reply nemo/collections/asr/modules/transformer_encoder.py]: Yeah actually, It is part of NeuralModule but since we not inheriting I added these methods separately for now.

Comment (pzelasko) [reply nemo/collections/asr/modules/transformer_encoder.py]: Maybe let's pull out the implementation from NeuralModule into free functions and import them both here and in NeuralModule? Call it "preparatory refactoring" :)

Comment (nithinraok) [reply nemo/collections/asr/modules/transformer_encoder.py]: Great suggestion. Applied the changes.

**Review by stevehuang52 — APPROVED** [review]
Looks good on my end, thanks!
